<a href="https://colab.research.google.com/github/chrisjallaine/CloudComputingDataEngineeringActivity/blob/main/toDataWarehouseFinal.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd
import numpy as np
import sqlalchemy
from sqlalchemy import create_engine, text
import time

### **Connect Database**

In [ ]:
# connect to database
conn = sqlalchemy.create_engine('postgresql://final_project_menorias_user:Ok2ZyjMuY33qLbpYmBTDTzLos6Gsrefa@dpg-d0shkf2li9vc73d0jpm0-a.oregon-postgres.render.com/final_project_menorias')


### **Extract Data from Database**

In [ ]:
# read cust data
cust = pd.read_sql('SELECT * FROM cust', conn)
cust

CID       BDATE     GEN
0      NASAW00011000  1971-10-06    Male
1      NASAW00011001  1976-05-10    Male
2      NASAW00011002  1971-02-09    Male
3      NASAW00011003  1973-08-14  Female
4      NASAW00011004  1979-08-05  Female
...              ...         ...     ...
18479     AW00029479  1969-06-30    None
18480     AW00029480  1977-05-06    None
18481     AW00029481  1965-07-04    None
18482     AW00029482  1964-09-01    None
18483     AW00029483  1965-06-06    None

[18484 rows x 3 columns]

In [ ]:
# read cust_info data
cust_info = pd.read_sql('SELECT * FROM cust_info', conn)
cust_info

cst_id     cst_key cst_firstname cst_lastname cst_marital_status  \
0      11000.0  AW00011000           Jon        Yang                   M   
1      11001.0  AW00011001        Eugene      Huang                    S   
2      11002.0  AW00011002         Ruben       Torres                  M   
3      11003.0  AW00011003       Christy          Zhu                  S   
4      11004.0  AW00011004     Elizabeth      Johnson                  S   
...        ...         ...           ...          ...                ...   
18489  29482.0  AW00029482       Clayton        Zhang                  M   
18490  29483.0  AW00029483          None      Navarro               None   
18491  29483.0  AW00029483          Marc      Navarro                  M   
18492      NaN    13451235          None         None               None   
18493      NaN      A01Ass          None         None               None   

      cst_gndr cst_create_date  
0            M      2025-10-06  
1            M      2025-10-06  
2            M      2025-10-06  
3            F      2025-10-06  
4            F      2025-10-06  
...        ...             ...  
18489     None      2026-01-25  
18490     None      2026-01-25  
18491     None      2026-01-27  
18492     None            None  
18493     None            None  

[18494 rows x 7 columns]

In [ ]:
# read loc data
loc= pd.read_sql('SELECT * FROM loc', conn)
loc

CID           CNTRY
0      AW-00011000       Australia
1      AW-00011001       Australia
2      AW-00011002       Australia
3      AW-00011003       Australia
4      AW-00011004       Australia
...            ...             ...
18479  AW-00029479          France
18480  AW-00029480  United Kingdom
18481  AW-00029481         Germany
18482  AW-00029482          France
18483  AW-00029483          France

[18484 rows x 2 columns]

In [ ]:
# read prd data
prd= pd.read_sql('SELECT * FROM prd', conn)
prd

prd_id           prd_key                     prd_nm  prd_cost prd_line  \
0       210  CO-RF-FR-R92B-58  HL Road Frame - Black- 58       NaN       R    
1       211  CO-RF-FR-R92R-58    HL Road Frame - Red- 58       NaN       R    
2       212   AC-HE-HL-U509-R      Sport-100 Helmet- Red      12.0       S    
3       213   AC-HE-HL-U509-R      Sport-100 Helmet- Red      14.0       S    
4       214   AC-HE-HL-U509-R      Sport-100 Helmet- Red      13.0       S    
..      ...               ...                        ...       ...      ...   
392     602     CO-BB-BB-8107          ML Bottom Bracket      45.0     None   
393     603     CO-BB-BB-9108          HL Bottom Bracket      54.0     None   
394     604  BI-RB-BK-R19B-44         Road-750 Black- 44     344.0       R    
395     605  BI-RB-BK-R19B-48         Road-750 Black- 48     344.0       R    
396     606  BI-RB-BK-R19B-52         Road-750 Black- 52     344.0       R    

    prd_start_dt  prd_end_dt  
0     2003-07-01        None  
1     2003-07-01        None  
2     2011-07-01  2007-12-28  
3     2012-07-01  2008-12-27  
4     2013-07-01        None  
..           ...         ...  
392   2013-07-01        None  
393   2013-07-01        None  
394   2013-07-01        None  
395   2013-07-01        None  
396   2013-07-01        None  

[397 rows x 7 columns]

In [ ]:
# read cat data
cat= pd.read_sql('SELECT * FROM cat', conn)
cat


ID          CAT             SUBCAT MAINTENANCE
0   AC_BR  Accessories         Bike Racks         Yes
1   AC_BS  Accessories        Bike Stands          No
2   AC_BC  Accessories  Bottles and Cages          No
3   AC_CL  Accessories           Cleaners         Yes
4   AC_FE  Accessories            Fenders          No
5   AC_HE  Accessories            Helmets         Yes
6   AC_HP  Accessories    Hydration Packs          No
7   AC_LI  Accessories             Lights         Yes
8   AC_LO  Accessories              Locks         Yes
9   AC_PA  Accessories           Panniers          No
10  AC_PU  Accessories              Pumps         Yes
11  AC_TT  Accessories    Tires and Tubes         Yes
12  BI_MB        Bikes     Mountain Bikes         Yes
13  BI_RB        Bikes         Road Bikes         Yes
14  BI_TB        Bikes      Touring Bikes         Yes
15  CL_BS     Clothing         Bib-Shorts          No
16  CL_CA     Clothing               Caps          No
17  CL_GL     Clothing             Gloves          No
18  CL_JE     Clothing            Jerseys          No
19  CL_SH     Clothing             Shorts          No
20  CL_SO     Clothing              Socks          No
21  CL_TI     Clothing             Tights          No
22  CL_VE     Clothing              Vests          No
23  CO_BB   Components    Bottom Brackets         Yes
24  CO_BR   Components             Brakes         Yes
25  CO_CH   Components             Chains         Yes
26  CO_CS   Components          Cranksets         Yes
27  CO_DE   Components        Derailleurs         Yes
28  CO_FO   Components              Forks         Yes
29  CO_HB   Components         Handlebars          No
30  CO_HS   Components           Headsets          No
31  CO_MF   Components    Mountain Frames         Yes
32  CO_PD   Components             Pedals          No
33  CO_RF   Components        Road Frames         Yes
34  CO_SA   Components            Saddles          No
35  CO_TF   Components     Touring Frames         Yes
36  CO_WH   Components             Wheels         Yes

In [ ]:
# read sales data
sales= pd.read_sql('SELECT * FROM sales', conn)
sales

sls_ord_num sls_prd_key  sls_cust_id  sls_order_dt  sls_ship_dt  \
0         SO43697  BK-R93R-62        21768      20101229     20110105   
1         SO43698  BK-M82S-44        28389      20101229     20110105   
2         SO43699  BK-M82S-44        25863      20101229     20110105   
3         SO43700  BK-R50B-62        14501      20101229     20110105   
4         SO43701  BK-M82S-44        11003      20101229     20110105   
...           ...         ...          ...           ...          ...   
60393     SO75122     FE-6654        15868      20140128     20140204   
60394     SO75122     CA-1098        15868      20140128     20140204   
60395     SO75123     FE-6654        18759      20140128     20140204   
60396     SO75123     ST-1401        18759      20140128     20140204   
60397     SO75123     CA-1098        18759      20140128     20140204   

       sls_due_dt  sls_sales  sls_quantity  sls_price  
0        20110110     3578.0             1     3578.0  
1        20110110     3400.0             1     3400.0  
2        20110110     3400.0             1     3400.0  
3        20110110      699.0             1      699.0  
4        20110110     3400.0             1     3400.0  
...           ...        ...           ...        ...  
60393    20140209       22.0             1       22.0  
60394    20140209        9.0             1        9.0  
60395    20140209       22.0             1       22.0  
60396    20140209      159.0             1      159.0  
60397    20140209        9.0             1        9.0  

[60398 rows x 9 columns]

### **Transformation: Merge Datasets**

This section is consisted of the merging process of the datasets between customer, product, and sales. The process involves cleaning on primary key columns (CID and prd_key) and then merge using join.

#### **Dataset 1: CUST_AZ12.csv**

In [ ]:
# remove the words in column CID just keep the numbers
cust['CID'] = cust['CID'].str.extract('(\d+)').astype(int)
cust

CID       BDATE     GEN
0      11000  1971-10-06    Male
1      11001  1976-05-10    Male
2      11002  1971-02-09    Male
3      11003  1973-08-14  Female
4      11004  1979-08-05  Female
...      ...         ...     ...
18479  29479  1969-06-30    None
18480  29480  1977-05-06    None
18481  29481  1965-07-04    None
18482  29482  1964-09-01    None
18483  29483  1965-06-06    None

[18484 rows x 3 columns]

#### **Dataset 2: cust_info.csv**

In [ ]:
# rename cst_key to CID
cust_info.rename(columns={'cst_key': 'CID'}, inplace=True)

In [ ]:
# remove the words in column CID just keep the numbers
cust_info['CID'] = cust_info['CID'].str.extract('(\d+)').astype(int)
cust_info

cst_id       CID cst_firstname cst_lastname cst_marital_status  \
0      11000.0     11000           Jon        Yang                   M   
1      11001.0     11001        Eugene      Huang                    S   
2      11002.0     11002         Ruben       Torres                  M   
3      11003.0     11003       Christy          Zhu                  S   
4      11004.0     11004     Elizabeth      Johnson                  S   
...        ...       ...           ...          ...                ...   
18489  29482.0     29482       Clayton        Zhang                  M   
18490  29483.0     29483          None      Navarro               None   
18491  29483.0     29483          Marc      Navarro                  M   
18492      NaN  13451235          None         None               None   
18493      NaN         1          None         None               None   

      cst_gndr cst_create_date  
0            M      2025-10-06  
1            M      2025-10-06  
2            M      2025-10-06  
3            F      2025-10-06  
4            F      2025-10-06  
...        ...             ...  
18489     None      2026-01-25  
18490     None      2026-01-25  
18491     None      2026-01-27  
18492     None            None  
18493     None            None  

[18494 rows x 7 columns]

#### **Dataset 3: LOC_A101.csv**

In [ ]:
# remove hyphen in 'CID' in loc df
loc['CID'] = loc['CID'].str.replace('-', '')
# remove the words in column CID just keep the numbers
loc['CID'] = loc['CID'].str.extract('(\d+)').astype(int)
loc

CID           CNTRY
0      11000       Australia
1      11001       Australia
2      11002       Australia
3      11003       Australia
4      11004       Australia
...      ...             ...
18479  29479          France
18480  29480  United Kingdom
18481  29481         Germany
18482  29482          France
18483  29483          France

[18484 rows x 2 columns]

#### **Merge the three dataset about the customers**

In [ ]:
# merge the two dataframes on CID - cust and cust_info
merged_df = pd.merge(cust, cust_info, on='CID', how='inner')
# display the merged dataframe
merged_df

CID       BDATE     GEN   cst_id cst_firstname cst_lastname  \
0      11000  1971-10-06    Male  11000.0           Jon        Yang    
1      11001  1976-05-10    Male  11001.0        Eugene      Huang     
2      11002  1971-02-09    Male  11002.0         Ruben       Torres   
3      11003  1973-08-14  Female  11003.0       Christy          Zhu   
4      11004  1979-08-05  Female  11004.0     Elizabeth      Johnson   
...      ...         ...     ...      ...           ...          ...   
18485  29480  1977-05-06    None  29480.0          Nina         Raji   
18486  29481  1965-07-04    None  29481.0          Ivan         Suri   
18487  29482  1964-09-01    None  29482.0       Clayton        Zhang   
18488  29483  1965-06-06    None  29483.0          None      Navarro   
18489  29483  1965-06-06    None  29483.0          Marc      Navarro   

      cst_marital_status cst_gndr cst_create_date  
0                      M        M      2025-10-06  
1                      S        M      2025-10-06  
2                      M        M      2025-10-06  
3                      S        F      2025-10-06  
4                      S        F      2025-10-06  
...                  ...      ...             ...  
18485                  S     None      2026-01-25  
18486                  S     None      2026-01-25  
18487                  M     None      2026-01-25  
18488               None     None      2026-01-25  
18489                  M     None      2026-01-27  

[18490 rows x 9 columns]

In [ ]:
# merge the two dataframes on CID - merged_df and loc
cust_df = pd.merge(merged_df, loc, on='CID', how='inner')
cust_df

CID       BDATE     GEN   cst_id cst_firstname cst_lastname  \
0      11000  1971-10-06    Male  11000.0           Jon        Yang    
1      11001  1976-05-10    Male  11001.0        Eugene      Huang     
2      11002  1971-02-09    Male  11002.0         Ruben       Torres   
3      11003  1973-08-14  Female  11003.0       Christy          Zhu   
4      11004  1979-08-05  Female  11004.0     Elizabeth      Johnson   
...      ...         ...     ...      ...           ...          ...   
18485  29480  1977-05-06    None  29480.0          Nina         Raji   
18486  29481  1965-07-04    None  29481.0          Ivan         Suri   
18487  29482  1964-09-01    None  29482.0       Clayton        Zhang   
18488  29483  1965-06-06    None  29483.0          None      Navarro   
18489  29483  1965-06-06    None  29483.0          Marc      Navarro   

      cst_marital_status cst_gndr cst_create_date           CNTRY  
0                      M        M      2025-10-06       Australia  
1                      S        M      2025-10-06       Australia  
2                      M        M      2025-10-06       Australia  
3                      S        F      2025-10-06       Australia  
4                      S        F      2025-10-06       Australia  
...                  ...      ...             ...             ...  
18485                  S     None      2026-01-25  United Kingdom  
18486                  S     None      2026-01-25         Germany  
18487                  M     None      2026-01-25          France  
18488               None     None      2026-01-25          France  
18489                  M     None      2026-01-27          France  

[18490 rows x 10 columns]

We have now the dataset of the customer details.

---

### **Dataset 4: prd_info.csv**

In [ ]:
# Create new column from PRD_ID by shortening to 4 characters and replacing '-' with '_'
prd['ID'] = prd['prd_key'].str[:5].str.replace('-', '_')
prd

prd_id           prd_key                     prd_nm  prd_cost prd_line  \
0       210  CO-RF-FR-R92B-58  HL Road Frame - Black- 58       NaN       R    
1       211  CO-RF-FR-R92R-58    HL Road Frame - Red- 58       NaN       R    
2       212   AC-HE-HL-U509-R      Sport-100 Helmet- Red      12.0       S    
3       213   AC-HE-HL-U509-R      Sport-100 Helmet- Red      14.0       S    
4       214   AC-HE-HL-U509-R      Sport-100 Helmet- Red      13.0       S    
..      ...               ...                        ...       ...      ...   
392     602     CO-BB-BB-8107          ML Bottom Bracket      45.0     None   
393     603     CO-BB-BB-9108          HL Bottom Bracket      54.0     None   
394     604  BI-RB-BK-R19B-44         Road-750 Black- 44     344.0       R    
395     605  BI-RB-BK-R19B-48         Road-750 Black- 48     344.0       R    
396     606  BI-RB-BK-R19B-52         Road-750 Black- 52     344.0       R    

    prd_start_dt  prd_end_dt     ID  
0     2003-07-01        None  CO_RF  
1     2003-07-01        None  CO_RF  
2     2011-07-01  2007-12-28  AC_HE  
3     2012-07-01  2008-12-27  AC_HE  
4     2013-07-01        None  AC_HE  
..           ...         ...    ...  
392   2013-07-01        None  CO_BB  
393   2013-07-01        None  CO_BB  
394   2013-07-01        None  BI_RB  
395   2013-07-01        None  BI_RB  
396   2013-07-01        None  BI_RB  

[397 rows x 8 columns]

#### **Dataset 5: PX_CAT_G1V2.csv**

In [ ]:
# mapping the category to prd df
prd_df = pd.merge(prd, cat, on='ID', how='left')
# merged dataframe with prd and cat
prd_df

prd_id           prd_key                     prd_nm  prd_cost prd_line  \
0       210  CO-RF-FR-R92B-58  HL Road Frame - Black- 58       NaN       R    
1       211  CO-RF-FR-R92R-58    HL Road Frame - Red- 58       NaN       R    
2       212   AC-HE-HL-U509-R      Sport-100 Helmet- Red      12.0       S    
3       213   AC-HE-HL-U509-R      Sport-100 Helmet- Red      14.0       S    
4       214   AC-HE-HL-U509-R      Sport-100 Helmet- Red      13.0       S    
..      ...               ...                        ...       ...      ...   
392     602     CO-BB-BB-8107          ML Bottom Bracket      45.0     None   
393     603     CO-BB-BB-9108          HL Bottom Bracket      54.0     None   
394     604  BI-RB-BK-R19B-44         Road-750 Black- 44     344.0       R    
395     605  BI-RB-BK-R19B-48         Road-750 Black- 48     344.0       R    
396     606  BI-RB-BK-R19B-52         Road-750 Black- 52     344.0       R    

    prd_start_dt  prd_end_dt     ID          CAT           SUBCAT MAINTENANCE  
0     2003-07-01        None  CO_RF   Components      Road Frames         Yes  
1     2003-07-01        None  CO_RF   Components      Road Frames         Yes  
2     2011-07-01  2007-12-28  AC_HE  Accessories          Helmets         Yes  
3     2012-07-01  2008-12-27  AC_HE  Accessories          Helmets         Yes  
4     2013-07-01        None  AC_HE  Accessories          Helmets         Yes  
..           ...         ...    ...          ...              ...         ...  
392   2013-07-01        None  CO_BB   Components  Bottom Brackets         Yes  
393   2013-07-01        None  CO_BB   Components  Bottom Brackets         Yes  
394   2013-07-01        None  BI_RB        Bikes       Road Bikes         Yes  
395   2013-07-01        None  BI_RB        Bikes       Road Bikes         Yes  
396   2013-07-01        None  BI_RB        Bikes       Road Bikes         Yes  

[397 rows x 11 columns]

In [ ]:
# remove the the first 6 characters from the 'prd_key' column
prd_df['prd_key'] = prd_df['prd_key'].str[6:]

#### **Merged two dataset about the prodcucts**

In [ ]:
prd_df

prd_id     prd_key                     prd_nm  prd_cost prd_line  \
0       210  FR-R92B-58  HL Road Frame - Black- 58       NaN       R    
1       211  FR-R92R-58    HL Road Frame - Red- 58       NaN       R    
2       212   HL-U509-R      Sport-100 Helmet- Red      12.0       S    
3       213   HL-U509-R      Sport-100 Helmet- Red      14.0       S    
4       214   HL-U509-R      Sport-100 Helmet- Red      13.0       S    
..      ...         ...                        ...       ...      ...   
392     602     BB-8107          ML Bottom Bracket      45.0     None   
393     603     BB-9108          HL Bottom Bracket      54.0     None   
394     604  BK-R19B-44         Road-750 Black- 44     344.0       R    
395     605  BK-R19B-48         Road-750 Black- 48     344.0       R    
396     606  BK-R19B-52         Road-750 Black- 52     344.0       R    

    prd_start_dt  prd_end_dt     ID          CAT           SUBCAT MAINTENANCE  
0     2003-07-01        None  CO_RF   Components      Road Frames         Yes  
1     2003-07-01        None  CO_RF   Components      Road Frames         Yes  
2     2011-07-01  2007-12-28  AC_HE  Accessories          Helmets         Yes  
3     2012-07-01  2008-12-27  AC_HE  Accessories          Helmets         Yes  
4     2013-07-01        None  AC_HE  Accessories          Helmets         Yes  
..           ...         ...    ...          ...              ...         ...  
392   2013-07-01        None  CO_BB   Components  Bottom Brackets         Yes  
393   2013-07-01        None  CO_BB   Components  Bottom Brackets         Yes  
394   2013-07-01        None  BI_RB        Bikes       Road Bikes         Yes  
395   2013-07-01        None  BI_RB        Bikes       Road Bikes         Yes  
396   2013-07-01        None  BI_RB        Bikes       Road Bikes         Yes  

[397 rows x 11 columns]

We have now the dataset pf the product details.

---

#### **Dataset 5: sales_details.csv**


In [ ]:
# rename the column names to match
sales.rename(columns={'sls_prd_key': 'prd_key', 'sls_cust_id': 'CID'}, inplace=True)
sales

sls_ord_num     prd_key    CID  sls_order_dt  sls_ship_dt  sls_due_dt  \
0         SO43697  BK-R93R-62  21768      20101229     20110105    20110110   
1         SO43698  BK-M82S-44  28389      20101229     20110105    20110110   
2         SO43699  BK-M82S-44  25863      20101229     20110105    20110110   
3         SO43700  BK-R50B-62  14501      20101229     20110105    20110110   
4         SO43701  BK-M82S-44  11003      20101229     20110105    20110110   
...           ...         ...    ...           ...          ...         ...   
60393     SO75122     FE-6654  15868      20140128     20140204    20140209   
60394     SO75122     CA-1098  15868      20140128     20140204    20140209   
60395     SO75123     FE-6654  18759      20140128     20140204    20140209   
60396     SO75123     ST-1401  18759      20140128     20140204    20140209   
60397     SO75123     CA-1098  18759      20140128     20140204    20140209   

       sls_sales  sls_quantity  sls_price  
0         3578.0             1     3578.0  
1         3400.0             1     3400.0  
2         3400.0             1     3400.0  
3          699.0             1      699.0  
4         3400.0             1     3400.0  
...          ...           ...        ...  
60393       22.0             1       22.0  
60394        9.0             1        9.0  
60395       22.0             1       22.0  
60396      159.0             1      159.0  
60397        9.0             1        9.0  

[60398 rows x 9 columns]

#### **Merge the customer and product dataframe to sales dataframe**

**Sales and Customer Dataframe**

In [ ]:
sales_df = sales.merge(cust_df, how='left', left_on='CID', right_on='CID', suffixes=('', '_cust'))
sales_df

sls_ord_num     prd_key    CID  sls_order_dt  sls_ship_dt  sls_due_dt  \
0         SO43697  BK-R93R-62  21768      20101229     20110105    20110110   
1         SO43698  BK-M82S-44  28389      20101229     20110105    20110110   
2         SO43699  BK-M82S-44  25863      20101229     20110105    20110110   
3         SO43700  BK-R50B-62  14501      20101229     20110105    20110110   
4         SO43701  BK-M82S-44  11003      20101229     20110105    20110110   
...           ...         ...    ...           ...          ...         ...   
60402     SO75122     FE-6654  15868      20140128     20140204    20140209   
60403     SO75122     CA-1098  15868      20140128     20140204    20140209   
60404     SO75123     FE-6654  18759      20140128     20140204    20140209   
60405     SO75123     ST-1401  18759      20140128     20140204    20140209   
60406     SO75123     CA-1098  18759      20140128     20140204    20140209   

       sls_sales  sls_quantity  sls_price       BDATE     GEN   cst_id  \
0         3578.0             1     3578.0  1952-02-19    None  21768.0   
1         3400.0             1     3400.0  1970-06-17  Female  28389.0   
2         3400.0             1     3400.0  1952-06-01  Female  25863.0   
3          699.0             1      699.0  1943-11-10    Male  14501.0   
4         3400.0             1     3400.0  1973-08-14  Female  11003.0   
...          ...           ...        ...         ...     ...      ...   
60402       22.0             1       22.0  1976-12-10    Male  15868.0   
60403        9.0             1        9.0  1976-12-10    Male  15868.0   
60404       22.0             1       22.0  1983-08-11    Male  18759.0   
60405      159.0             1      159.0  1983-08-11    Male  18759.0   
60406        9.0             1        9.0  1983-08-11    Male  18759.0   

      cst_firstname cst_lastname cst_marital_status cst_gndr cst_create_date  \
0              Cole       Watson                  S        M      2026-01-05   
1           Rachael     Martinez                  S     None      2026-01-25   
2            Sydney       Wright                  S        F      2026-01-14   
3             Ruben       Prasad                  M     None      2025-10-12   
4           Christy          Zhu                  S        F      2025-10-06   
...             ...          ...                ...      ...             ...   
60402         Caleb          Lal                  S        M      2026-01-04   
60403         Caleb          Lal                  S        M      2026-01-04   
60404         Devin     Phillips                  S        M      2026-01-04   
60405         Devin     Phillips                  S        M      2026-01-04   
60406         Devin     Phillips                  S        M      2026-01-04   

               CNTRY  
0             Canada  
1             France  
2      United States  
3                USA  
4          Australia  
...              ...  
60402         Canada  
60403         Canada  
60404         Canada  
60405         Canada  
60406         Canada  

[60407 rows x 18 columns]

**Merge with Products**

In [ ]:
# merge the two dataframes on prd_key
sales_df = sales_df.merge(prd_df, how='left', left_on='prd_key', right_on='prd_key', suffixes=('', '_prd'))
sales_df

sls_ord_num     prd_key    CID  sls_order_dt  sls_ship_dt  sls_due_dt  \
0         SO43697  BK-R93R-62  21768      20101229     20110105    20110110   
1         SO43698  BK-M82S-44  28389      20101229     20110105    20110110   
2         SO43699  BK-M82S-44  25863      20101229     20110105    20110110   
3         SO43700  BK-R50B-62  14501      20101229     20110105    20110110   
4         SO43700  BK-R50B-62  14501      20101229     20110105    20110110   
...           ...         ...    ...           ...          ...         ...   
89828     SO75123     FE-6654  18759      20140128     20140204    20140209   
89829     SO75123     ST-1401  18759      20140128     20140204    20140209   
89830     SO75123     CA-1098  18759      20140128     20140204    20140209   
89831     SO75123     CA-1098  18759      20140128     20140204    20140209   
89832     SO75123     CA-1098  18759      20140128     20140204    20140209   

       sls_sales  sls_quantity  sls_price       BDATE  ... prd_id  \
0         3578.0             1     3578.0  1952-02-19  ...    310   
1         3400.0             1     3400.0  1970-06-17  ...    346   
2         3400.0             1     3400.0  1952-06-01  ...    346   
3          699.0             1      699.0  1943-11-10  ...    336   
4          699.0             1      699.0  1943-11-10  ...    337   
...          ...           ...        ...         ...  ...    ...   
89828       22.0             1       22.0  1983-08-11  ...    485   
89829      159.0             1      159.0  1983-08-11  ...    486   
89830        9.0             1        9.0  1983-08-11  ...    223   
89831        9.0             1        9.0  1983-08-11  ...    224   
89832        9.0             1        9.0  1983-08-11  ...    225   

                        prd_nm prd_cost prd_line prd_start_dt  prd_end_dt  \
0             Road-150 Red- 62   2171.0       R    2011-07-01  2007-12-28   
1      Mountain-100 Silver- 44   1912.0       M    2011-07-01  2007-12-28   
2      Mountain-100 Silver- 44   1912.0       M    2011-07-01  2007-12-28   
3           Road-650 Black- 62    413.0       R    2011-07-01  2007-12-28   
4           Road-650 Black- 62    487.0       R    2012-07-01  2008-12-27   
...                        ...      ...      ...          ...         ...   
89828    Fender Set - Mountain      8.0       M    2013-07-01        None   
89829   All-Purpose Bike Stand     59.0       M    2013-07-01        None   
89830             AWC Logo Cap      6.0       S    2011-07-01  2007-12-28   
89831             AWC Logo Cap      5.0       S    2012-07-01  2008-12-27   
89832             AWC Logo Cap      7.0       S    2013-07-01        None   

          ID          CAT          SUBCAT MAINTENANCE  
0      BI_RB        Bikes      Road Bikes         Yes  
1      BI_MB        Bikes  Mountain Bikes         Yes  
2      BI_MB        Bikes  Mountain Bikes         Yes  
3      BI_RB        Bikes      Road Bikes         Yes  
4      BI_RB        Bikes      Road Bikes         Yes  
...      ...          ...             ...         ...  
89828  AC_FE  Accessories         Fenders          No  
89829  AC_BS  Accessories     Bike Stands          No  
89830  CL_CA     Clothing            Caps          No  
89831  CL_CA     Clothing            Caps          No  
89832  CL_CA     Clothing            Caps          No  

[89833 rows x 28 columns]

We have now one dataset `sales_df` that is consisted of customers, products, and sales data.

---

### **Transformation: Data Cleaning and Preprocessing**

In [ ]:
data = sales_df

In [ ]:
data.head()

sls_ord_num     prd_key    CID  sls_order_dt  sls_ship_dt  sls_due_dt  \
0     SO43697  BK-R93R-62  21768      20101229     20110105    20110110   
1     SO43698  BK-M82S-44  28389      20101229     20110105    20110110   
2     SO43699  BK-M82S-44  25863      20101229     20110105    20110110   
3     SO43700  BK-R50B-62  14501      20101229     20110105    20110110   
4     SO43700  BK-R50B-62  14501      20101229     20110105    20110110   

   sls_sales  sls_quantity  sls_price       BDATE  ... prd_id  \
0     3578.0             1     3578.0  1952-02-19  ...    310   
1     3400.0             1     3400.0  1970-06-17  ...    346   
2     3400.0             1     3400.0  1952-06-01  ...    346   
3      699.0             1      699.0  1943-11-10  ...    336   
4      699.0             1      699.0  1943-11-10  ...    337   

                    prd_nm prd_cost prd_line prd_start_dt  prd_end_dt     ID  \
0         Road-150 Red- 62   2171.0       R    2011-07-01  2007-12-28  BI_RB   
1  Mountain-100 Silver- 44   1912.0       M    2011-07-01  2007-12-28  BI_MB   
2  Mountain-100 Silver- 44   1912.0       M    2011-07-01  2007-12-28  BI_MB   
3       Road-650 Black- 62    413.0       R    2011-07-01  2007-12-28  BI_RB   
4       Road-650 Black- 62    487.0       R    2012-07-01  2008-12-27  BI_RB   

     CAT          SUBCAT MAINTENANCE  
0  Bikes      Road Bikes         Yes  
1  Bikes  Mountain Bikes         Yes  
2  Bikes  Mountain Bikes         Yes  
3  Bikes      Road Bikes         Yes  
4  Bikes      Road Bikes         Yes  

[5 rows x 28 columns]

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 89833 entries, 0 to 89832
Data columns (total 28 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   sls_ord_num         89833 non-null  object 
 1   prd_key             89833 non-null  object 
 2   CID                 89833 non-null  int64  
 3   sls_order_dt        89833 non-null  int64  
 4   sls_ship_dt         89833 non-null  int64  
 5   sls_due_dt          89833 non-null  int64  
 6   sls_sales           89818 non-null  float64
 7   sls_quantity        89833 non-null  int64  
 8   sls_price           89822 non-null  float64
 9   BDATE               89833 non-null  object 
 10  GEN                 83450 non-null  object 
 11  cst_id              89833 non-null  float64
 12  cst_firstname       89823 non-null  object 
 13  cst_lastname        89822 non-null  object 
 14  cst_marital_status  89826 non-null  object 
 15  cst_gndr            58500 non-null  object 
 16  cst_

# Dataset Metadata

| Column Name            | Non-Null Count | Data Type | Description (Inferred)                     | Notes                              |
|------------------------|----------------|-----------|--------------------------------------------|------------------------------------|
| **sls_ord_num**        | 89,833         | `object`  | Sales order number (unique identifier)     | Likely alphanumeric.               |
| **prd_key**           | 89,833         | `object`  | Product key (identifier)                   | Foreign key for products.          |
| **CID**               | 89,833         | `int64`   | Customer ID                                | Unique customer identifier.        |
| **sls_order_dt**      | 89,833         | `int64`   | Order date (numeric format)                | Needs conversion to datetime.      |
| **sls_ship_dt**       | 89,833         | `int64`   | Shipment date (numeric format)             | Needs conversion to datetime.      |
| **sls_due_dt**        | 89,833         | `int64`   | Due date (numeric format)                  | Needs conversion to datetime.      |
| **sls_sales**         | 89,818         | `float64` | Total sales amount                         | 15 missing values.                 |
| **sls_quantity**      | 89,833         | `int64`   | Quantity of products sold                  |                                    |
| **sls_price**         | 89,822         | `float64` | Unit price of product                      | 11 missing values.                 |
| **BDATE**             | 89,833         | `object`  | Birthdate (string format)                  | Needs datetime parsing.            |
| **GEN**               | 83,450         | `object`  | Generation (e.g., Gen X, Millennial)       | 6,383 missing values.              |
| **cst_id**            | 89,833         | `float64` | Customer ID (alternative)                  | Duplicate of `CID`?                |
| **cst_firstname**     | 89,823         | `object`  | Customer first name                        | 10 missing values.                 |
| **cst_lastname**      | 89,822         | `object`  | Customer last name                         | 11 missing values.                 |
| **cst_marital_status**| 89,826         | `object`  | Marital status (e.g., Single, Married)     | 7 missing values.                  |
| **cst_gndr**          | 58,500         | `object`  | Customer gender                            | 31,333 missing values (35%).       |
| **cst_create_date**   | 89,833         | `object`  | Customer account creation date             | Needs datetime parsing.            |
| **CNTRY**             | 88,484         | `object`  | Customer country                           | 1,349 missing values.              |
| **prd_id**            | 89,833         | `int64`   | Product ID                                 | Unique identifier.                 |
| **prd_nm**            | 89,833         | `object`  | Product name                               |                                    |
| **prd_cost**          | 89,833         | `float64` | Product cost                               |                                    |
| **prd_line**          | 89,833         | `object`  | Product line/category                      | e.g., Electronics, Apparel.        |
| **prd_start_dt**      | 89,833         | `object`  | Product launch date                        | Needs datetime parsing.            |
| **prd_end_dt**        | 32,667         | `object`  | Product discontinuation date               | 57,166 missing (likely active).    |
| **ID**                | 89,833         | `object`  | Additional identifier (ambiguous)          | Check business context.            |
| **CAT**               | 89,833         | `object`  | Product category                           | e.g., Hardware, Software.          |
| **SUBCAT**            | 89,833         | `object`  | Product subcategory                        |                                    |
| **MAINTENANCE**       | 89,833         | `object`  | Maintenance flag (e.g., Yes/No)            | Binary or descriptive?             |

### Key Observations:
1. **Temporal Columns**:  
   - Dates (`sls_order_dt`, `sls_ship_dt`, `BDATE`, etc.) are stored as `int64` or `object`—need conversion to datetime.
2. **Missing Values**:  
   - `cst_gndr` (35% missing), `prd_end_dt` (64% missing), `GEN` (7% missing).
3. **Potential Duplicates**:  
   - `CID` (int) vs. `cst_id` (float)—verify if they represent the same entity.
4. **Product Lifecycle**:  
   - `prd_end_dt` missing for most entries, suggesting active products.
5. **Sales Metrics**:  
   - `sls_sales`, `sls_price` have minimal missing values (<0.02%).

### Data Cleaning

In [26]:
#  Fix and rename the messy date columns
def fix_messy_date(val):
    try:
        val_str = str(int(val))
        if len(val_str) == 8:
            return pd.to_datetime(val_str, format='%Y%m%d', errors='coerce').date()
    except:
        return pd.NaT
    return pd.NaT

messy_date_map = {
    'sls_order_dt': 'order_date',
    'sls_ship_dt': 'ship_date',
    'sls_due_dt': 'due_date'
}

for old_col, new_col in messy_date_map.items():
    if old_col in data.columns:
        data[new_col] = data[old_col].apply(fix_messy_date)
        data.drop(old_col, axis=1, inplace=True)
    else:
        print(f"Column '{old_col}' not found — skipping.")

#  Clean conversion for already well-formed date columns
clean_columns_map = {
    'BDATE': 'birth_date',
    'cst_create_date': 'customer_creation_date',
    'prd_start_dt': 'product_start_date',
    'prd_end_dt': 'product_end_date'
}

for old_col, new_col in clean_columns_map.items():
    if old_col in data.columns:
        data[new_col] = pd.to_datetime(data[old_col], errors='coerce').dt.date
        data.drop(old_col, axis=1, inplace=True)
    else:
        print(f"Column '{old_col}' not found — skipping.")

#### Handle Missing Values

In [28]:
data['is_active_product'] = data['product_end_date'].isna().astype(int)

In [29]:
data['is_active_product'] = data['product_end_date'].isna().astype(int)

In [31]:
# Low Missing Values

numeric_cols = ['sls_sales', 'sls_price', 'prd_cost']
data[numeric_cols] = data[numeric_cols].fillna(data[numeric_cols].median())

categorical_cols = ['cst_firstname', 'cst_lastname', 'cst_marital_status']
data[categorical_cols] = data[categorical_cols].fillna('Not Specified')

#### Standardize Categorical Data

In [32]:
data['GEN'] = data['GEN'].str.upper().str.strip()
data['CNTRY'] = data['CNTRY'].str.upper().str.strip()

In [33]:
gender_map = {'Male': 'M', 'Female': 'F', 'm': 'M', 'f': 'F'}
data['cst_gndr'] = data['cst_gndr'].replace(gender_map).fillna('Unknown')

#### De-duplicate and validate identifiers

In [34]:
if data['CID'].equals(data['cst_id'].astype('int64')):
    data.drop('cst_id', axis=1, inplace=True)  # Remove duplicate

In [35]:
duplicate_count = len(data) - data['sls_ord_num'].nunique()
print(f"{duplicate_count} duplicate order numbers found.")

62174 duplicate order numbers found.


In [36]:
data = data.drop_duplicates(subset='sls_ord_num', keep='first')

#### Renaming

In [37]:
data['cst_marital_status'].unique()

array(['S', 'M', 'Not Specified'], dtype=object)

#### Handle Product Data

In [38]:
data['full_category'] = data['CAT'] + ' > ' + data['SUBCAT']

<ipython-input-38-57c9986c7931>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data['full_category'] = data['CAT'] + ' > ' + data['SUBCAT']


In [40]:
from scipy import stats
data = data[(np.abs(stats.zscore(data[['sls_quantity', 'sls_price']])) < 3).all(axis=1)]

#### Feature Engineering

In [41]:
# Customer Age

# Convert birth_date to pandas datetime (Timestamp)
data['birth_date'] = pd.to_datetime(data['birth_date'], errors='coerce')

# Calculate age in years (floor division)
data['customer_age'] = (pd.to_datetime('today') - data['birth_date']).dt.days // 365

In [42]:
data['customer_age'] = data['customer_age'].astype('Int64')

In [43]:
# Order-to-Ship Delay
print(data[['order_date', 'ship_date']].dtypes)

order_date    object
ship_date     object
dtype: object


In [44]:
for col in ['order_date', 'ship_date']:
    data[col] = pd.to_datetime(data[col], errors='coerce')

In [45]:
data['order_to_ship_days'] = (data['ship_date'] - data['order_date']).dt.days

#### Drop Unecessary Columns

In [46]:
data.columns

Index(['sls_ord_num', 'prd_key', 'CID', 'sls_sales', 'sls_quantity',
       'sls_price', 'GEN', 'cst_firstname', 'cst_lastname',
       'cst_marital_status', 'cst_gndr', 'CNTRY', 'prd_id', 'prd_nm',
       'prd_cost', 'prd_line', 'ID', 'CAT', 'SUBCAT', 'MAINTENANCE',
       'order_date', 'ship_date', 'due_date', 'birth_date',
       'customer_creation_date', 'product_start_date', 'product_end_date',
       'is_active_product', 'full_category', 'customer_age',
       'order_to_ship_days'],
      dtype='object')

In [47]:
data.drop(columns=[
    'full_category', 'cst_gndr'
], inplace=True)

#### Validation

In [48]:
data.isna().sum()

sls_ord_num                   0
prd_key                       0
CID                           0
sls_sales                     0
sls_quantity                  0
sls_price                     0
GEN                        1820
cst_firstname                 0
cst_lastname                  0
cst_marital_status            0
CNTRY                       332
prd_id                        0
prd_nm                        0
prd_cost                      0
prd_line                      0
ID                            0
CAT                           0
SUBCAT                        0
MAINTENANCE                   0
order_date                    8
ship_date                     0
due_date                      0
birth_date                    9
customer_creation_date        0
product_start_date            0
product_end_date          17087
is_active_product             0
customer_age                  9
order_to_ship_days            8
dtype: int64

In [49]:
data = data.dropna()

In [50]:
data['order_to_ship_days'] = data['order_to_ship_days'].astype('Int64')

In [51]:
for col in ['due_date', 'customer_creation_date', 'product_start_date', 'product_end_date']:
    data[col] = pd.to_datetime(data[col], errors='coerce').dt.date

In [52]:
data['is_active_product'] = data['is_active_product'].astype(bool)

In [53]:
for col in ['order_date', 'ship_date', 'due_date']:
    data[col] = pd.to_datetime(data[col], errors='coerce').dt.date


In [54]:
date_cols = [
    'order_date', 'ship_date', 'due_date',
    'birth_date', 'customer_creation_date',
    'product_start_date', 'product_end_date'
]

for col in date_cols:
    data[col] = pd.to_datetime(data[col], errors='coerce')

In [55]:
for col in date_cols:
    data[f'{col}_month'] = data[col].dt.month_name()

In [56]:
data.head()

sls_ord_num     prd_key    CID  sls_sales  sls_quantity  sls_price     GEN  \
1     SO43698  BK-M82S-44  28389     3400.0             1     3400.0  FEMALE   
2     SO43699  BK-M82S-44  25863     3400.0             1     3400.0  FEMALE   
3     SO43700  BK-R50B-62  14501      699.0             1      699.0    MALE   
5     SO43701  BK-M82S-44  11003     3400.0             1     3400.0  FEMALE   
6     SO43702  BK-R93R-44  27645     3578.0             1     3578.0    MALE   

  cst_firstname cst_lastname cst_marital_status  ... is_active_product  \
1       Rachael     Martinez                  S  ...             False   
2        Sydney       Wright                  S  ...             False   
3         Ruben       Prasad                  M  ...             False   
5       Christy          Zhu                  S  ...             False   
6         Colin        Anand                  M  ...             False   

   customer_age order_to_ship_days  order_date_month ship_date_month  \
1            54                  7          December         January   
2            73                  7          December         January   
3            81                  7          December         January   
5            51                  7          December         January   
6            52                  7          December         January   

  due_date_month birth_date_month customer_creation_date_month  \
1        January             June                      January   
2        January             June                      January   
3        January         November                      October   
5        January           August                      October   
6        January           August                      January   

  product_start_date_month product_end_date_month  
1                     July               December  
2                     July               December  
3                     July               December  
5                     July               December  
6                     July               December  

[5 rows x 36 columns]

In [57]:
# Export to CSV without the DataFrame index (this is jsut for checking)
data.to_csv('dataset.csv', index=False)

### Send To Warehouse

In [58]:
!pip install pandas sqlalchemy psycopg2-binary

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.0/3.0 MB 59.6 MB/s eta 0:00:00


In [59]:
#Render database details (kiy render)
DATABASE_URL = "postgresql://user:szrVeYFcCzLxNRLlxxPtX6w4fWqM1ulr@dpg-d0oun3muk2gs73945erg-a.singapore-postgres.render.com/salesdashboard_of09"

In [60]:
# Create a SQLAlchemy engine
engine = create_engine(DATABASE_URL)

# Export DataFrame to PostgreSQL
data.to_sql(
    name='sales_database',
    con=engine,
    if_exists='replace',
    index=False,
    chunksize=1000
)

print("Data successfully exported to Render PostgreSQL!")

Data successfully exported to Render PostgreSQL!


In [61]:
# Read back data to confirm
df_loaded = pd.read_sql('SELECT * FROM sales_database LIMIT 10', con=engine)
df_loaded

sls_ord_num     prd_key    CID  sls_sales  sls_quantity  sls_price     GEN  \
0     SO43698  BK-M82S-44  28389     3400.0             1     3400.0  FEMALE   
1     SO43699  BK-M82S-44  25863     3400.0             1     3400.0  FEMALE   
2     SO43700  BK-R50B-62  14501      699.0             1      699.0    MALE   
3     SO43701  BK-M82S-44  11003     3400.0             1     3400.0  FEMALE   
4     SO43702  BK-R93R-44  27645     3578.0             1     3578.0    MALE   
5     SO43703  BK-R93R-62  16624     3578.0             1     3578.0    MALE   
6     SO43704  BK-M82B-48  11005     3375.0             1     3375.0    MALE   
7     SO43705  BK-M82S-38  11011     3400.0             1     3400.0    MALE   
8     SO43706  BK-R93R-48  27621     3578.0             1     3578.0    MALE   
9     SO43707  BK-R93R-48  27616     3578.0             1     3578.0  FEMALE   

  cst_firstname cst_lastname cst_marital_status  ... is_active_product  \
0       Rachael     Martinez                  S  ...             False   
1        Sydney       Wright                  S  ...             False   
2         Ruben       Prasad                  M  ...             False   
3       Christy          Zhu                  S  ...             False   
4         Colin        Anand                  M  ...             False   
5        Albert      Alvarez                  S  ...             False   
6         Julio         Ruiz                  S  ...             False   
7        Curtis           Lu                  M  ...             False   
8        Edward        Brown                  S  ...             False   
9          Emma        Brown                  S  ...             False   

   customer_age order_to_ship_days  order_date_month ship_date_month  \
0            54                  7          December         January   
1            73                  7          December         January   
2            81                  7          December         January   
3            51                  7          December         January   
4            52                  7          December         January   
5            41                  7          December         January   
6            48                  7          December         January   
7            56                  7          December         January   
8            50                  7          December         January   
9            54                  7          December         January   

  due_date_month birth_date_month customer_creation_date_month  \
0        January             June                      January   
1        January             June                      January   
2        January         November                      October   
3        January           August                      October   
4        January           August                      January   
5        January             July                      January   
6        January           August                      October   
7        January              May                      October   
8        January          January                      January   
9        January             June                      January   

  product_start_date_month product_end_date_month  
0                     July               December  
1                     July               December  
2                     July               December  
3                     July               December  
4                     July               December  
5                     July               December  
6                     July               December  
7                     July               December  
8                     July               December  
9                     July               December  

[10 rows x 36 columns]